### DISCUSSION TITLE
How to list allowed Content Type collections from within the API controller

### DISCUSSION THREAD

**Discussion by Pilskalns** [strapi/strapi#7165] [Legacy Topics]
I need to create an API point that would list available Content-Types for the authenticated user. If permission configuration changes for the Role or new content type is created, then it should reflect in the content.

On most basic output it would work something like

` GET /`
```
{
    "content-endpoints": [
        "/news",
        "/products",
        "/case-studies",
        "/etc"
    ]
}
```
Presumably, it should be based on whether the `{content type}.find` route is allowed.

I've successfully created a controller for `GET /`, but I cannot figure out how to compile the list. From the[ Global API](https://strapi.io/documentation/v3.x/global-strapi/api-reference.html#strapi-api) `strapi.api` it is possible to find `apiName` inside models etc. but it does not tell whether current user can execute `news.find` or `products.find`...

Pls halp 🙃!

**Comment by lauriejim** [comment]
Hello! It's quite simple because this code already exist 😝
It's what we are doing to check if you can or can't do requisition (permissions check)

Please check the docs of routing and controller to see how create a custom route.
After that, in your code:
Here is how to get the current user role - `const role = ctx.state.user.role`
And to get permissions of this role
```js
const permissions = await strapi.query('permission', 'users-permissions').find({
    role: role.id
  });
```

**Reply by Pilskalns** [reply]
After hours and hours of reading code and debugging, I just figured it out on my own which turns out is the same method as in your example. Great, it means I am on the right way to do this.

Except for one very tiny but !important detail - in the route controller, before calling find query I had to use `next()` callback. Otherwise, the `.user` was not populated in the context, which was my pitfall all the time.

So, in brevity, the controller now looks like this:

```node
module.exports = {
    find: async (ctx, next) => {
      next(); // !important - makes this to execute after permissions are settled

    // some error checking here whether the request is properly authorized and have a user

      const permissions = await strapi.query('permission', 'users-permissions').find(
        {
          role: ctx.state.user.role.id,
          type: 'application',
          action: 'find',
          enabled: true,
        },
        []
      );
    // format permissions into nice return object, then ctx.send()
  }
};
```

**Reply by lauriejim** [reply]
Hum that's weird .... How did you create your stuff?
Because you should not have to do that.

**Reply by Pilskalns** [reply]
I created a new project from scratch, added modifications and it works without the need of `next()`.
Probably it was due to bad permission records in DB - I installed and removed my code as a plugin, as a controller in different configuration ways many times.
